README.md
<p align="center">
  <img alt="react-native-swiper" src="http://i.imgur.com/P4cRUgD.png" width="208">
</p>

<p align="center">
  The best Swiper component for React Native.
</p>

<p align="center">
  <a href="http://standardjs.com/"><img alt="JavaScript Style Guide" src="https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/react-native-swiper"><img alt="npm version" src="http://img.shields.io/npm/v/react-native-swiper.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/react-native-swiper"><img alt="npm version" src="http://img.shields.io/npm/dm/react-native-swiper.svg?style=flat-square"></a>
  <a href="https://github.com/leecade/react-native-swiper/pulls?q=is%3Apr+is%3Aclosed"><img alt="PR Stats" src="https://img.shields.io/issuestats/i/github/leecade/react-native-swiper.svg?style=flat-square"></a>
  <a href="https://github.com/leecade/react-native-swiper/issues?q=is%3Aissue+is%3Aclosed"><img alt="Issue Stats" src="https://img.shields.io/issuestats/p/github/leecade/react-native-swiper.svg?style=flat-square"></a>
  <a href="https://gitter.im/leecade/react-native-swiper?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge"><img alt="Join the chat" src="https://badges.gitter.im/leecade/react-native-swiper.svg"></a>
</p>

# react-native-swiper

## Roadmap

> see: [ROADMAP.md](ROADMAP.md)

## Changelogs

- **[1.6.0]**
  + Refactor and tidy up to simplify the code.
  + Change pagination to use the visual position of the swiper instead of the final resting position - keeps things looking much more fluid.
  + *Breaking - Separate Buttons and Pagination into separate files, removed pass through of style props through main Swiper - new method is to pass through a styled version of these components or generate your own 
  + *Breaking - Remove title, buggy and wasn't configurable

- **[1.5.4]**
  + Added easily accessible pagination point manipulation: use `dotColor` / `activeDotColor` and `dotStyle` / `activeDotStyle` (thanks [@denizs](https://github.com/denizs))
  + Added scrollEnabled prop to documentation (thanks [@ibandominguez](https://github.com/ibandominguez))

- [1.5.3]
  + Add loadMinimalLoader prop to customize `<ActivityIndicator />` (thanks [@Exilz](https://github.com/Exilz))
  + Disable autoplay timer when prop changes to false (thanks [@dizlexik](https://github.com/dizlexik))
  + Special thanks to [@hypatiah](https://github.com/dizlexik) for fixed some grammatical errors in README

- [1.5.2]
  + Add yarn lock
  + Fix jitter when quickly swiping back and forth between pages (iOS) (thanks [@nemophrost](https://github.com/nemophrost))
  + The first webview always reloaded when injecting the rest of the children (thanks [@eosterberg](https://github.com/eosterberg))

> see more: [CHANGELOG.md](CHANGELOG.md)

## Show Cases

> Try these cases by yourself very easy, Just open `examples/ios/swiper.xcodeproj` in Xcode, then press `Cmd + R`; you may edit `examples/index.ios.js` for switch cases.

### [examples/components/Basic](https://github.com/leecade/react-native-swiper/blob/master/examples/components/Basic)

![](http://i.imgur.com/zrsazAG.gif =300x)

### [examples/components/Swiper](https://github.com/leecade/react-native-swiper/blob/master/examples/components/Swiper)

![](http://i.imgur.com/hP3f3oO.gif =300x)

### [examples/components/SwiperNumber](https://github.com/leecade/react-native-swiper/blob/master/examples/components/SwiperNumber)

![](http://i.imgur.com/0rqESVb.gif =300x)

### [examples/components/Phone](https://github.com/leecade/react-native-swiper/blob/master/examples/components/Phone)

![](http://i.imgur.com/c1BhjZm.gif =300x)

### [examples/components/LoadMinimal](https://github.com/leecade/react-native-swiper/blob/master/examples/components/LoadMinimal)

![](http://i.imgur.com/LAOHbJA.gif =300x)

### [examples/components/PhotoView](https://github.com/leecade/react-native-swiper/blob/master/examples/components/PhotoView)

![](http://i.imgur.com/GkIRzjO.gif =300x)

> with [react-native-photo-view](https://github.com/alwx/react-native-photo-view)

## Getting Started

- [Installation](#installation)
- [Basic Usage](#basic-usage)
- [Properties](#properties)
  + [Basic](#basic)
  + [Custom basic style & content](#custom-basic-style--content)
  + [Pagination](#pagination)
  + [Autoplay](#autoplay)
  + [Control buttons](#control-buttons)
  + [Props of Children](#props-of-children)
  + [Basic props of `<ScrollView />`](#basic-props-of-scrollview-)
  + [Supported ScrollResponder](#supported-scrollresponder)
- [Examples](#examples)
- [Development](#development)

### Installation

```bash
$ npm i react-native-swiper --save
```

### Basic Usage

- Install `react-native` first

```bash
$ npm i react-native -g
```

- Initialization of a react-native project

```bash
$ react-native init myproject
```

- Then, edit `myproject/index.ios.js`, like this:

```jsx
import React, { Component } from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  View
} from 'react-native';

import Swiper from 'react-native-swiper';

var styles = StyleSheet.create({
  wrapper: {
  },
  slide1: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#9DD6EB',
  },
  slide2: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#97CAE5',
  },
  slide3: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#92BBD9',
  },
  text: {
    color: '#fff',
    fontSize: 30,
    fontWeight: 'bold',
  }
})

var swiper = React.createClass({
  render: function() {
    return (
      <Swiper style={styles.wrapper} showsButtons={true}>
        <View style={styles.slide1}>
          <Text style={styles.text}>Hello Swiper</Text>
        </View>
        <View style={styles.slide2}>
          <Text style={styles.text}>Beautiful</Text>
        </View>
        <View style={styles.slide3}>
          <Text style={styles.text}>And simple</Text>
        </View>
      </Swiper>
    )
  }
})

AppRegistry.registerComponent('myproject', () => swiper);
```

### Properties

#### Basic

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| horizontal | true | `bool` | If `true`, the scroll view's children are arranged horizontally in a row instead of vertically in a column. |
| loop | true | `bool` | Set to `false` to disable continuous loop mode. |
| index | 0 | `number` | Index number of initial slide. |
| showsButtons | false | `bool` | Set to `true` make control buttons visible. |
| autoplay | false | `bool` | Set to `true` enable auto play mode. |

#### Custom basic style & content

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| width | - | `number` | If no specify default enable fullscreen mode by `flex: 1`. |
| height | - | `number` | If no specify default fullscreen mode by `flex: 1`. |
| style | {...} | `style` | See default style in source. |
| loadMinimal | false | `bool` | Only load current index slide , `loadMinimalSize` slides before and after. |
| loadMinimalSize | 1 | `number` | see `loadMinimal`   |
| loadMinimalLoader | `<ActivityIndicator />` | `element` | Custom loader to display when slides aren't loaded

#### Pagination

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| showsPagination | true | `bool` | Set to `true` make pagination visible. |
| paginationStyle | {...} | `style` | Custom styles will merge with the default styles. |
| renderPagination | - | `function` | Complete control how to render pagination with three params (`index`, `total`, `context`) ref to `this.state.index` / `this.state.total` / `this`, For example: show numbers instead of dots. |
| dot | `<View style={{backgroundColor:'rgba(0,0,0,.2)', width: 8, height: 8,borderRadius: 4, marginLeft: 3, marginRight: 3, marginTop: 3, marginBottom: 3,}} />` | `element` | Allow custom the dot element. |
| activeDot | `<View style={{backgroundColor: '#007aff', width: 8, height: 8, borderRadius: 4, marginLeft: 3, marginRight: 3, marginTop: 3, marginBottom: 3,}} />` | `element` | Allow custom the active-dot element. |
| dotStyle | - | `object` | Allow custom the active-dot element. |
| dotColor | - | `string` | Allow custom the active-dot element. |
| activeDotColor | - | `string` | Allow custom the active-dot element. |
| activeDotStyle | - | `object` | Allow custom the active-dot element. |

#### Autoplay

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| autoplay | true | `bool` | Set to `true` enable auto play mode. |
| autoplayTimeout | 2.5 | `number` | Delay between auto play transitions (in second). |
| autoplayDirection | true | `bool` | Cycle direction control. |

#### Control buttons

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| showsButtons | true | `bool` | Set to `true` make control buttons visible. |
| buttonWrapperStyle | `{backgroundColor: 'transparent', flexDirection: 'row', position: 'absolute', top: 0, left: 0, flex: 1, paddingHorizontal: 10, paddingVertical: 10, justifyContent: 'space-between', alignItems: 'center'}` | `style` | Custom styles. |
| nextButton | `<Text style={styles.buttonText}>›</Text>` | `element` | Allow custom the next button. |
| prevButton | `<Text style={styles.buttonText}>‹</Text>` | `element` | Allow custom the prev button. |

#### Props of Children

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| style | {...} | `style` | Custom styles will merge with the default styles. |
| title | {<Text numberOfLines={1}>...</Text>} | `element` | If this parameter is not specified, will not render the title. |

#### Basic props of `<ScrollView />`

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| horizontal | true | `bool` | If `true`, the scroll view's children are arranged horizontally in a row instead of vertically in a column. |
| pagingEnabled | true | `bool` | If true, the scroll view stops on multiples of the scroll view's size when scrolling. This can be used for horizontal pagination.  |
| showsHorizontalScrollIndicator | false | `bool` | Set to `true` if you want to show horizontal scroll bar. |
| showsVerticalScrollIndicator | false | `bool` |  Set to `true` if you want to show vertical scroll bar. |
| bounces | false | `bool` | If `true`, the scroll view bounces when it reaches the end of the content if the content is larger then the scroll view along the axis of the scroll direction. If `false`, it disables all bouncing even if the alwaysBounce* props are true.  |
| scrollsToTop | false | `bool` | If true, the scroll view scrolls to top when the status bar is tapped.  |
| removeClippedSubviews | true | `bool` | If true, offscreen child views (whose overflow value is hidden) are removed from their native backing superview when offscreen. This canimprove scrolling performance on long lists.  |
| automaticallyAdjustContentInsets | false | `bool` | Set to `true` if you need adjust content insets automation. |
| scrollEnabled | true | `bool` | Enables/Disables swiping |

> @see: http://facebook.github.io/react-native/docs/scrollview.html

#### Supported ScrollResponder

| Prop  | Params  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| onScrollBeginDrag | `e` / `state` / `context` | `function` | When animation begins after letting up |
| onMomentumScrollEnd | `e` / `state` / `context` | `function` | Makes no sense why this occurs first during bounce |
| onTouchStartCapture | `e` / `state` / `context` | `function` | Immediately after `onMomentumScrollEnd` |
| onTouchStart | `e` / `state` / `context` | `function` | Same, but bubble phase |
| onTouchEnd | `e` / `state` / `context` | `function` | You could hold the touch start for a long time |
| onResponderRelease | `e` / `state` / `context` | `function` | When lifting up - you could pause forever before * lifting |

> Note: each ScrollResponder be injected with two params: `state` and `context`, you can get `state` and `context`(ref to swiper's `this`) from params, for example:

```jsx
var swiper = React.createClass({
  _onMomentumScrollEnd: function (e, state, context) {
    console.log(state, context.state)
  },
  render: function() {
    return (
      <Swiper style={styles.wrapper}
      onMomentumScrollEnd ={this._onMomentumScrollEnd}
     ...
      </Swiper>
    )
  }
})
```

> More ScrollResponder info, see: https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollResponder.js

### Methods

#### scrollBy(index, animated)

Scroll by relative index.

Parameters:

| Name  | Type     | default | Description |
| :---- | :------: | :------: | :--- |
| index | `number`   | `undefined` | offset index |
| animated | `bool`   | `true` | offset index |

### Examples

```bash
$ cd examples
$ npm i
$ react-native run-ios
```

> Quick start with [examples](https://github.com/leecade/react-native-swiper/tree/master/examples/).

### Development

```bash
$ cd examples
$ npm i
$ npm run dev
$ react-native run-ios
```

Then launch simulator to preview. Note that you just need to edit the source file `src/index.js`, the change will auto sync to examples.

And now that this project follows the [standard](https://github.com/feross/standard) code style, you'd better prepare it for IDE.

## Contribution

- [@leecade](mailto:leecade@163.com) The main author.
- [@rajkissu](mailto:rajkissu@gmail.com) The secondary contributor.

## Questions

Feel free to [contact me](mailto:leecade@163.com) or [create an issue](https://github.com/leecade/react-native-swiper/issues/new)

> Inspired by [nolimits4web/Swiper](https://github.com/nolimits4web/swiper/) & Design material from [Dribbble](https://dribbble.com/) & made with ♥.


CHANGELOG.md
- [1.6.0]
  + Refactor and tidy up to simplify the code.
  + Change pagination to use the visual position of the swiper instead of the final resting position - keeps things looking much more fluid.
  + *Breaking - Separate Buttons and Pagination into separate files, removed pass through of style props through main Swiper - new method is to pass through a styled version of these components or generate your own 
  + *Breaking - Remove title, buggy and wasn't configurable

- [1.5.1]
  + Allow scroll without animate, ref: [scrollBy(index, animated)](#scrollbyindex-animated)
  + Remove [#254](https://github.com/leecade/react-native-swiper/pull/254) which break the scroll direction in loop mode

- [1.5.0]
  + Improve develop [workflow](#development) for support watch & use real pkg name in examples
  + Update examples for support `React-native 0.33` + `Xcode8` + `ios10` + `iphone7`
  + Switch to [standard](https://github.com/feross/standard) code style
  + Rewrite source code with ES6 syntex
  + Cancel transform the source code when publish
  + Add a [demo](examples/components/LoadMinimal) for `loadMinimal` + `lazyload`
  + Add a [demo](examples/components/PhotoView) for multi-images viewer
  + [#254](https://github.com/leecade/react-native-swiper/pull/254) Fix jitter when quickly swiping back and forth between pages (iOS) (@thanks [@nemophrost](https://github.com/nemophrost)).

- [1.4.11]
  + Adds loadMinimalSize property

- [1.4.10]
  + Adds loadMinimal to api (lazyloading-esque)

- [1.4.9]
  + Adds extra check for title

- [1.4.8]
  + Fixes loop and autoplay bugs in android

- [1.4.7]
  + Don't attempt to mutate this.props
  + Fixes examples links
  + Adds drag end handling to always reset state.isScrolling
  + Fixes float calculation error

- [1.4.6]
  + refactors examples
  + prevents mutation of `this.props`
  + fixes android index loop issue

- [1.4.5]
  + renames `scrollTo()` to `scrollBy()`
  + image index is now always an integer
  + prevents parent state updates from reseting index counter
  + fixes issue with scrolling not working sometimes

- [1.4.4]
  + Support for React Native latest(`^0.26.0`)
  + Updates examples to work with react-native-swiper 1.4.4
  + Examples now run on Android (some are still buggy, needs more work)
  + Removes old examples folder

- [1.4.3]
  + Fixed auto play issue when only a child view

- [v1.4.0]
  + Support for React Native latest(`^0.17.0`) & Upgrade examples base on Xcode `7.2` / React Native `0.17.0`

  + Upgrade build tool to babel6

  + Add missing deps `react-timer-mixin`

  + [1f8643a](https://github.com/leecade/react-native-swiper/commit/1f8643a67e2768d165132a19629a991a86672036) Move Dimensions import to deconstructed React import for react-native (@thanks [@jedrekk](https://github.com/jedrekk)).

  + [e28af9b](https://github.com/leecade/react-native-swiper/commit/e28af9b205f17447cb3149b45fc220beec037bce) Support index property in combination with loop property (@thanks [@almost](https://github.com/almost)).

  + [6c832d](https://github.com/leecade/react-native-swiper/commit/6c832d6a23da3737a2e8a2667273dc6093bcc9ee) fix warnings about keys for dots (@thanks [@sunnylqm](https://github.com/sunnylqm)).

  + [8de1afc](https://github.com/leecade/react-native-swiper/commit/8de1afcb75a003424231bb089802db53bbbf84e4) Changes to make the example work in xcode 7 (@thanks [@allomov](https://github.com/allomov)).

- [v1.3.0]
  + [8d6d75c](https://github.com/leecade/react-native-swiper/commit/8d6d75c00edf87b603c218aad0018932277814b5) inject `state` in ScrollResponders (@thanks [@smothers](https://github.com/smothers)).

- [v1.2.2]
  + [890c0ce](https://github.com/leecade/react-native-swiper/commit/890c0ce64e2192c2ca7830e6699f67b88171e74b) ensure `onMomentumScrollEnd` synchronous update `index`.

- [v1.2.0]
  + [838b24c](https://github.com/leecade/react-native-swiper/commit/838b24cbeaf49b9ca1dabb4eed8305e314503fb1) Re-design a loop model for avoid img flickering.
  + [9cb91c5](https://github.com/leecade/react-native-swiper/commit/9cb91c58c84034b0b8b874dbfc2a44da982686a8) fixes [#7](https://github.com/leecade/react-native-swiper/issues/6) `onMomentumScrollEnd` lost `event` param. (thanks [@smothers](https://github.com/smothers))

- [v1.1.1]
  + [21f0b00](https://github.com/leecade/react-native-swiper/commit/21f0b00138b6936cd3dfac2eb107a14c99c7392b) fixes [#6](https://github.com/leecade/react-native-swiper/issues/6) a define-propType error. (thanks [@benjamingeorge](https://github.com/benjamingeorge))

- [v1.1.0]
  + [44ec630](https://github.com/leecade/react-native-swiper/commit/44ec630b62844dbeaccee73adaa0996e319ebffb) fixes [#4](https://github.com/leecade/react-native-swiper/issues/4) `onMoementumScrollEnd` gets overridden. (thanks [@subpublicanders](https://github.com/subpublicanders))
  + [5de06a7](https://github.com/leecade/react-native-swiper/commit/5de06a7aa86318ad38720728022b80e5cf98a2ab) New prop: `renderPagination`. (thanks [@aksonov](https://github.com/aksonov))

- [v1.0.4]
  + [21cb373](https://github.com/leecade/react-native-swiper/commit/21cb3732578588f9d47ee7ddda541577ad691970) fixes [#2](https://github.com/leecade/react-native-swiper/issues/2) Solve the problem of installation. (thanks [@jamwaffles](https://github.com/jamwaffles))

- [v1.0.3]
  + [0f796f3](https://github.com/leecade/react-native-swiper/commit/0f796f3557b5aeb1772573cd7ecae2e835bccc0b) fixes [#1](https://github.com/leecade/react-native-swiper/issues/1) Two 'horizontal' in propTypes. (thanks [@MacyzZ](https://github.com/MacyzZ))


src/Buttons.js
/**
 * react-native-swiper
 * @author leecade<leecade@163.com>
 */

import React, {PropTypes} from 'react';
import {View, Text, TouchableOpacity, StyleSheet} from 'react-native';

export const Buttons = ({index, total, loop, scrollBackward, scrollForward, buttonWrapperStyle}) => {
  const showNextButton = loop || index !== total - 1;
  const showPrevButton = loop || index !== 0;

  return (
    <View pointerEvents="box-none" style={[styles.buttonWrapper, buttonWrapperStyle]}>
      {showPrevButton ? (
          <TouchableOpacity onPress={scrollBackward}>
            <View>
              <Text style={styles.buttonText}>‹</Text>
            </View>
          </TouchableOpacity>
        ) : <View /> }
      {showNextButton ? (
          <TouchableOpacity onPress={scrollForward}>
            <View>
              <Text style={styles.buttonText}>›</Text>
            </View>
          </TouchableOpacity>
        ) : <View /> }
    </View>
  );
};

Buttons.propTypes = {
  total: PropTypes.number.isRequired,
  index: PropTypes.number.isRequired,

  scrollBackward: PropTypes.func,
  scrollForward: PropTypes.func,

  loop: PropTypes.bool,
  buttonWrapperStyle: View.propTypes.style
};

const styles = StyleSheet.create({
  buttonWrapper: {
    backgroundColor: 'transparent',
    flexDirection: 'row',
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    paddingHorizontal: 10,
    paddingVertical: 10,
    justifyContent: 'space-between',
    alignItems: 'center'
  },

  buttonText: {
    fontSize: 50,
    color: '#007aff',
    fontFamily: 'Arial'
  },
});


src/Swiper.js
/* global setImmediate */

/**
 * react-native-swiper
 * @author leecade<leecade@163.com>
 */
import React, {Component, PropTypes} from 'react';
import {
  View,
  ScrollView,
  Dimensions,
  ViewPagerAndroid,
  Platform,
  ActivityIndicator,
  StyleSheet,
} from 'react-native';

import {Pagination} from './Pagination';
import {Buttons} from './Buttons';

const {width, height} = Dimensions.get('window');

class Swiper extends Component {
  /**
   * Props Validation
   * @type {Object}
   */
  static propTypes = {
    ...ScrollView.propTypes,

    showsPagination: PropTypes.bool,
    renderPagination: PropTypes.func,
    showsButtons: PropTypes.bool,
    renderButtons: PropTypes.func,

    loadMinimal: PropTypes.bool,
    loadMinimalSize: PropTypes.number,
    loadMinimalLoader: PropTypes.element,
    loop: PropTypes.bool,
    autoPlay: PropTypes.bool,
    autoPlayTimeout: PropTypes.number,
    autoPlayDirection: PropTypes.bool,
    index: PropTypes.number,
  };

  /**
   * Default props
   * @return {object} props
   * @see http://facebook.github.io/react-native/docs/scrollview.html
   */
  static defaultProps = {
    index: 0,
    horizontal: true,
    pagingEnabled: true,
    showsHorizontalScrollIndicator: false,
    showsVerticalScrollIndicator: false,
    bounces: false,
    scrollsToTop: false,
    removeClippedSubviews: true,
    automaticallyAdjustContentInsets: false,
    showsPagination: true,
    showsButtons: false,
    loop: true,
    loadMinimal: false,
    loadMinimalSize: 1,
    autoPlay: false,
    autoPlayTimeout: 2.5,
    autoPlayDirection: true,
    scrollEventThrottle: 16,
  };

  /**
   * Init states
   * @return {object} states
   */
  state = this.initState(this.props, true);

  componentWillReceiveProps(nextProps) {
    const sizeChanged = (nextProps.width || width) !== this.state.width ||
      (nextProps.height || height) !== this.state.height;

    if (!nextProps.autoPlay) {
      clearTimeout(this.autoPlayTimer);
    }
    this.setState(this.initState(nextProps, sizeChanged));
  }

  componentDidMount() {
    this.autoPlay();
  }

  componentWillUnmount() {
    clearTimeout(this.autoPlayTimer);
    clearTimeout(this.loopJumpTimer);
  }

  autoPlayTimer = null;
  loopJumpTimer = null;
  offset = 0;
  isScrolling = false;

  initState(props, sizeChanged) {
    // set the current state
    const state = this.state || {};

    this.isScrolling = false;
    this.offset = 0;

    const total = props.children.length || 1;

    const totalChanged = state.total !== total;
    const boundedIndex = total > 1 ? Math.min(props.index, total - 1) : 0;
    const index = totalChanged ? boundedIndex : state.index;

    // Default: horizontal
    const newWidth = props.width || width;
    const newHeight = props.height || height;

    if (total > 1) {
      const startIndex = index + (props.loop ? 1 : 0);
      const step = props.horizontal ? newWidth : newHeight;
      this.offset = startIndex * step;
    }

    return {
      total,
      index,
      visualIndex: state.visualIndex || 0,
      autoPlayEnd: false,
      loopJump: false,
      width: newWidth,
      height: newHeight,
      // only update the offset in state if needed, updating offset while swiping
      // causes some bad jumping / stuttering
      offset: totalChanged || sizeChanged ? this.offset : state.offset,
    };
  }

  // include private vars with state
  fullState() {
    return Object.assign({}, this.state, {isScrolling: this.isScrolling, offset: this.offset});
  }

  loopJump() {
    const {loop} = this.props;
    const {loopJump, index} = this.state;

    if (!loopJump) {
      return;
    }

    if (this.scrollView && this.scrollView.setPageWithoutAnimation) {
      const i = index + (loop ? 1 : 0);
      this.loopJumpTimer = setTimeout(() => this.scrollView.setPageWithoutAnimation(i), 50);
    }
  }

  /**
   * Automatic rolling
   */
  autoPlay() {
    const {total, index, autoPlayEnd} = this.state;
    const {autoPlay, loop, autoPlayTimeout, autoPlayDirection} = this.props;

    if (this.isScrolling || total < 2 || !autoPlay || autoPlayEnd) {
      return;
    }

    clearTimeout(this.autoPlayTimer);

    this.autoPlayTimer = setTimeout(() => {
      if (!loop && (autoPlayDirection ? index === total - 1 : index === 0)) {
        this.setState({autoPlayEnd: true});
      } else {
        this.scrollBy(autoPlayDirection ? 1 : -1);
      }
    }, autoPlayTimeout * 1000);
  }

  /**
   * Scroll begin handle
   * @param  {object} e native event
   */
  onScrollBegin = (e) => {
    const {onScrollBeginDrag} = this.props;

    this.isScrolling = true;

    onScrollBeginDrag && onScrollBeginDrag(e, this.fullState(), this);
  };

  /**
   * Scroll end handle
   * @param  {object} e native event
   */
  onScrollEnd = (e) => {
    const {nativeEvent: {contentOffset, position}} = e;
    const {width, height} = this.state;
    const {onMomentumScrollEnd, horizontal} = this.props;

    this.isScrolling = false;

    let modifiedContentOffset;
    // making our events coming from android compatible to updateIndex logic
    if (contentOffset) {
      modifiedContentOffset = horizontal ? contentOffset.x : contentOffset.y;
    } else {
      modifiedContentOffset = horizontal ? position * width : position * height;
    }

    this.updateIndex(modifiedContentOffset, () => {
      this.autoPlay();
      this.loopJump();

      // if `onMomentumScrollEnd` registered will be called here
      onMomentumScrollEnd && onMomentumScrollEnd(e, this.fullState(), this);
    });
  };

  /*
   * Drag end handle
   * @param {object} e native event
   */
  onScrollEndDrag = ({nativeEvent: {contentOffset}}) => {
    const {horizontal} = this.props;
    const {index, total} = this.state;
    const previousOffset = this.offset;
    const newOffset = horizontal ? contentOffset.x : contentOffset.y;

    if (previousOffset === newOffset && (index === 0 || index === total - 1)) {
      this.isScrolling = false;
    }
  };

  /*
   * Handle onScroll events and maintain the visualIndex
   * @param {object} e native event
   */
  onScroll = ({nativeEvent: {contentOffset}}) => {
    const {loop, horizontal} = this.props;
    const {width, height, total, visualIndex} = this.state;
    const step = horizontal ? width : height;
    const pos = horizontal ? contentOffset.x : contentOffset.y;

    let newVisualIndex = Math.floor((pos + (step / 2)) / step);

    // Looping index is offset by 1, needs to be bounded
    if (loop) {
      newVisualIndex--;

      if (newVisualIndex < 0) {
        newVisualIndex = total - 1;
      } else if (newVisualIndex >= total) {
        newVisualIndex = 0;
      }
    }

    if (newVisualIndex !== visualIndex) {
      this.setState({visualIndex: newVisualIndex});
    }
  };

  /**
   * Update index after scroll
   * @param  {object} offset content offset
   * @param  {func} callback
   */
  updateIndex(offset, cb) {
    const {index, width, height, total} = this.state;
    const {loop, horizontal} = this.props;

    const diff = offset - this.offset;

    // Do nothing if offset no change.
    if (!diff) {
      return;
    }

    const step = horizontal ? width : height;

    // Note: if touch very very quickly and continuous,
    // the variation of `index` more than 1.
    // parseInt() ensures it's always an integer
    let newIndex = parseInt(index + Math.round(diff / step), 10);
    let newOffset = offset;
    let loopJump = false;

    if (loop) {
      if (newIndex <= -1) {
        newIndex = total - 1;
        newOffset = step * total;
        loopJump = true;
      } else if (newIndex >= total) {
        newIndex = 0;
        newOffset = step;
        loopJump = true;
      }
    }

    this.offset = newOffset;

    const newState = {
      index: newIndex,
      visualIndex: newIndex,
      loopJump,
    };

    // only update offset in state if loopJump is true
    if (loopJump) {
      // when swiping to the beginning of a looping set for the third time,
      // the new offset will be the same as the last one set in state.
      // Setting the offset to the same thing will not do anything,
      // so we increment it by 1 then immediately set it to what it should be,
      // after render.
      if (newOffset === this.state.offset) {
        newState.offset = newOffset + 1;
        this.setState(newState, () => {
          this.setState({offset: newOffset}, cb);
        });
      } else {
        newState.offset = newOffset;
        this.setState(newState, cb);
      }
    } else {
      this.setState(newState, cb);
    }
  }

  /**
   * Scroll by index
   * @param  {number} indexOffset offset index
   * @param  {bool} animated
   */

  scrollBy = (indexOffset, animated = true) => {
    const {horizontal, loop} = this.props;
    const {index, total} = this.state;

    if (this.isScrolling || total < 2) {
      return;
    }

    const state = this.state;
    const diff = (loop ? 1 : 0) + indexOffset + index;
    const x = horizontal ? diff * state.width : 0;
    const y = horizontal ? 0 : diff * state.height;

    if (Platform.OS === 'android') {
      this.scrollView && this.scrollView[animated ? 'setPage' : 'setPageWithoutAnimation'](diff);
    } else {
      this.scrollView && this.scrollView.scrollTo({x, y, animated});
    }

    // update scroll state
    this.isScrolling = true;

    this.setState({autoPlayEnd: false});

    // trigger onScrollEnd manually in android
    if (!animated || Platform.OS === 'android') {
      setImmediate(() => this.onScrollEnd({nativeEvent: {position: diff}}));
    }
  };

  scrollForward = () => this.scrollBy(1);
  scrollBackward = () => this.scrollBy(-1);

  scrollViewPropOverrides() {
    /*
     const scrollResponders = [
     'onMomentumScrollBegin',
     'onTouchStartCapture',
     'onTouchStart',
     'onTouchEnd',
     'onResponderRelease',
     ]
     */
    return Object.entries(this.props).reduce((overrides, [key, value]) => {
      if (typeof value === 'function' &&
        key !== 'onMomentumScrollEnd' &&
        key !== 'renderPagination' &&
        key !== 'onScrollBeginDrag'
      ) {
        overrides[key] = (e) => value(e, this.fullState(), this);
      }
    }, {});
  }

  /**
   * Render pagination
   * @return {object} react-dom
   */
  renderPagination() {
    const {total, index, visualIndex} = this.state;
    const {showsPagination, renderPagination} = this.props;

    if (showsPagination) {
      if (renderPagination) {
        return renderPagination({index, visualIndex, total, context: this});
      }

      return <Pagination index={visualIndex} total={total} />;
    }
  }

  renderButtons() {
    const {loop, showsButtons, renderButtons} = this.props;
    const {index, total} = this.state;

    if (showsButtons) {
      const buttonProps = {index, total, loop, scrollBackward: this.scrollBackward, scrollForward: this.scrollForward};
      if (renderButtons) {
        return renderButtons({...buttonProps, context: this});
      }
      return <Buttons {...buttonProps} />;
    }
  }

  buildPages() {
    const {width, height, index, total} = this.state;
    const {children, loop, loadMinimal, loadMinimalSize, loadMinimalLoader} = this.props;

    const loopVal = loop ? 1 : 0;

    const dimensions = {width, height};
    const pageStyle = [dimensions, styles.slide];
    const pageStyleLoading = [dimensions, styles.pageLoading];

    if (total <= 1) {
      return [<View style={pageStyle} key={0}>{children}</View>];
    }

    // Re-design a loop model for avoid img flickering
    const pageKey = Object.keys(children);
    if (loop) {
      pageKey.unshift(String(total - 1));
      pageKey.push('0');
    }

    return pageKey.map((page, i) => {
      if (loadMinimal) {
        if (i >= (index + loopVal - loadMinimalSize) &&
          i <= (index + loopVal + loadMinimalSize)) {
          return <View style={pageStyle} key={i}>{children[page]}</View>;
        }
        return (
          <View style={pageStyleLoading} key={`loading-${i}`}>
            {loadMinimalLoader ? loadMinimalLoader : <ActivityIndicator />}
          </View>
        );
      }

      return <View style={pageStyle} key={i}>{children[page]}</View>;
    });
  }

  setScrollView = (element) => {
    this.scrollView = element;
  };

  renderScrollView() {
    const pages = this.buildPages();

    const {style, ...props} = this.props;
    const {offset, index} = this.state;

    const contentOffset = {
      x: this.props.horizontal ? offset : 0,
      y: this.props.horizontal ? 0 : offset,
    };

    if (Platform.OS === 'ios') {
      return (
        <ScrollView ref={this.setScrollView}
                    {...props}
                    {...this.scrollViewPropOverrides()}
                    contentContainerStyle={[styles.wrapper, style]}
                    contentOffset={contentOffset}
                    onScrollBeginDrag={this.onScrollBegin}
                    onMomentumScrollEnd={this.onScrollEnd}
                    onScrollEndDrag={this.onScrollEndDrag}
                    onScroll={this.onScroll}>
          {pages}
        </ScrollView>
      );
    }
    return (
      <ViewPagerAndroid ref="scrollView"
                        {...this.props}
                        initialPage={this.props.loop ? index + 1 : index}
                        onPageSelected={this.onScrollEnd}
                        style={{flex: 1}}>
        {pages}
      </ViewPagerAndroid>
    );
  }

  /**
   * Default render
   * @return {object} react node
   */
  render() {
    const {width, height} = this.state;

    return (
      <View style={[styles.container, {width, height}]}>
        {this.renderScrollView()}
        {this.renderPagination()}
        {this.renderButtons()}
      </View>
    );
  }
}

export default Swiper;

/**
 * Default styles
 * @type {StyleSheetPropType}
 */
const styles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
    position: 'relative'
  },

  wrapper: {
    backgroundColor: 'transparent'
  },

  slide: {
    backgroundColor: 'transparent'
  },

  pageLoading: {
    justifyContent: 'center',
    alignItems: 'center'
  },
});
